src/MICore/TerminalLauncher.cs

```diff
@@ -0,0 +1,152 @@
+                terminal = new LinuxTerminalLauncher(title, initScript, environment);
+            }
+            else if (PlatformUtilities.IsOSX())
+            {
+                terminal = new MacTerminalLauncher(title, initScript, environment);
+            }
+            else
+            {
+                Debug.Fail("Cannot make a terminal for non Linux or OS X.");
+                throw new InvalidOperationException();
+            }
+
+            return terminal;
+        }
+
+        protected abstract string GetProcessExecutable();
+        protected abstract string GetProcessArgs();
+
+        protected TerminalLauncher(string title, string initScript, ReadOnlyCollection<EnvironmentEntry> environment)
+        {
+            _title = title;
+            _initScript = initScript;
+            _environment = environment;
+            _isRoot = UnixNativeMethods.GetEUid() == 0;
+        }
+
+        public void Launch(string workingDirectory)
+        {
+            _terminalProcess = new Process
+            {
+                StartInfo =
+                {
+                    CreateNoWindow = false,
+                    UseShellExecute = false,
+                    WorkingDirectory = workingDirectory,
+                    FileName = GetProcessExecutable(),
+                    Arguments = GetProcessArgs()
+                }
+            };
+
+            CopyEnvironment();
+            _terminalProcess.Start();
+        }
+
+        private void CopyEnvironment()
+        {
+            if (_environment == null)
+            {
+                return;
+            }
+
+            ProcessStartInfo processStartInfo = _terminalProcess.StartInfo;
+            foreach (EnvironmentEntry entry in _environment)
+            {
+                processStartInfo.SetEnvironmentVariable(entry.Name, entry.Value);
+            }
+        }
+    }
+
+    internal class MacTerminalLauncher : TerminalLauncher
+    {
+        private const string OsascriptPath = "/usr/bin/osascript";
+        private const string LaunchTerminalScript = "osxlaunchhelper.scpt";
+
+        public MacTerminalLauncher(string title, string initScript, ReadOnlyCollection<EnvironmentEntry> environment)
+            : base(title, initScript, environment)
+        { }
+
+        protected override string GetProcessExecutable()
+        {
+            return OsascriptPath;
+        }
+
+        protected override string GetProcessArgs()
+        {
+            string thisModulePath = typeof(MacTerminalLauncher).GetTypeInfo().Assembly.ManifestModule.FullyQualifiedName;
+            string launchScript = Path.Combine(Path.GetDirectoryName(thisModulePath), LaunchTerminalScript);
+            if (!File.Exists(launchScript))
+            {
+                throw new FileNotFoundException(MICoreResources.Error_InternalFileMissing);
+            }
+
+            return string.Format(CultureInfo.InvariantCulture, "{0} \"{1}\" \"{2}\"", launchScript, _title, _initScript);
```

Comment:
Why?

---
In reply to: [61978609](https://github.com/Microsoft/MIEngine/pull/303#discussion_r61978609) [](ancestors = 61978609)